### DISCUSSION TITLE
多重高级字幕并载

### DISCUSSION THREAD

**Discussion by hooke007** [hooke007/mpv_PlayKit#61] [INFO 信息]
# 本文已过时

参见 https://github.com/hooke007/MPV_lazy/discussions/61#discussioncomment-7888729

<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>

## 前置说明

_当我发现这个功能就在我眼皮底下时我是又无语又兴奋的_  
_在mpv真正实现完美的双字幕前这可能是最佳的解决方案，虽然使用方式很”程序员“。。。_

- mpv中滤镜的具体使用说明看 [文档](https://hooke007.github.io/unofficial/mpv_filters.html) ，这里不详述。

> 当前mpv自带实现的双字幕功能的局限 https://mpv.io/manual/master/#options-secondary-sid  
这个需求的最大受众需要的是并列两条高级字幕（比如一边看弹幕一边看外挂特效字幕），然而此时次字幕（如果是文本类高级字幕）只能以纯文本的形式在画面顶部输出（但图形字幕不受此限制）

## 过去的尝试

我之前测试双字幕用的是 [VSFilterMod](https://github.com/Masaiki/VSFilterMod) /  [AssRender](https://github.com/AmusementClub/assrender) /  [xy-VSFilter](https://github.com/pinterf/xy-VSFilter)，以VapourSynth为媒介参与渲染。效果乍一看没毛病，但是受限于 [video_in](https://mpv.io/manual/master/#video-filters-video-in) ，无法进行时间轴对齐（一旦进行跳转操作，vs的字幕立刻从0开始重新输出）。

## 可行的办法

后来我发现lavfi下的subtitles滤镜就能粗糙地实现这个需求，随意跳转时间轴完全没有进度重置的问题，甚至能并列输出多重高级字幕。。。

_通过堆叠三个subtitles滤镜，加上mpv原生输出，一共四条字幕同时渲染没有大问题_
![多重高级字幕](https://user-images.githubusercontent.com/41094733/145698981-a825ad43-e4d0-47e2-9736-790e6a18bf77.jpg)

### 操作流程

1. mpv加载文件后呼出控制台
2. 输入最简化的命令例如 `vf toggle subtitles=字幕.ass` 即可成功装载字幕
_控制台支持中文（外部复制完文本后直接ctrl+v粘贴）_
除了手册03里提及的查询方式，该滤镜所有可用的子参数还可以看 **[这里](https://ffmpeg.org/ffmpeg-all.html#subtitles-1)**

### 明显的局限

- [ ] 滤镜特性需要强制软解或auto-copy
- [x] <s>视频内置的字幕无法作为次字幕输出</s>
 可以指向同文件时用子参数 `stream_index` 指定内置的目标字幕轨道（从0计数），示例指定源视频的内置第三字幕轨道：
```
vf toggle ``subtitles=filename="D:\路径\视频.mkv":stream_index=2``
```
- [x] <s>敲命令很不优雅（如果你的目标字幕不在当前工作目录，需要使用绝对路径指定）</s>
 不完全的解决方案 https://github.com/hooke007/MPV_lazy/discussions/61#discussioncomment-1842073
- [ ] 次字幕本身就有时间偏差怎么办 —— 凉拌

**Comment by hooke007** [comment]
<s>[open_dialog.lua](https://github.com/hooke007/MPV_lazy/blob/7124c926533771c9dc3c8208a79abe93251ffe59/portable_config/scripts/open_dialog.lua) </s> **[load_plus.lua](https://github.com/hooke007/MPV_lazy/blob/main/portable_config/scripts/load_plus.lua)** 弥补了上述方案中只能敲命令的缺陷，按照示例设置文件部署快捷键即可激活对应功能。

因为使用了标签，在变更为次字幕为其它文件时，依旧保持了其在滤镜列表中的位置。
如果不想要这样不符合append直觉的特性，你需要更换次字幕文件时先移除再加载。

_该脚本只支持挂载一路次字幕，不考虑更高的数量需求_

**Reply by dyphire** [reply]
很骚的操作，是否考虑实现常规的加载指定外挂字幕/音轨的功能，类似mpv.net那样

**Reply by dyphire** [reply]
修改了下open_dialog.lua脚本实现了加载指定外挂字幕/音轨的功能，已pr

**Reply by wathper** [reply]
‘示例设置文件部署快捷键’指的是这一行代码吗？
ALT+w    script-binding load_plus/append_sid     # 追加其它字幕（切换）

**Reply by hooke007** [reply]
https://github.com/hooke007/MPV_lazy/blob/ec01ac4ea2bd58b9a86220aa0aee5a8c98b1ea47/portable_config/input_scripts.conf#L37-L39

**Reply by wathper** [reply]
原来是这个，我以为是可以调节次字幕时轴的没有用过的特性。上面写的时间偏差的介绍我看漏了。

**Reply by ipcjs** [reply]
脚本只能读取视频文件中的第一个字幕，要怎么再加个弹窗选择字幕？🤔️

**Reply by hooke007** [reply]
不支持

**Reply by ipcjs** [reply]
借助ChatGPT，土法糊了一个😅：
https://gist.github.com/ipcjs/e568793fc7f06155e5db7320c7ca328d

**Comment by dyphire** [comment]
通过个人修改版 [mpv](https://github.com/dyphire/mpv) 添加了 [--secondary-sub-override](https://github.com/dyphire/mpv/commit/e3300e025e86fc4c2643c1f5fcfa1e9c637bc413) 选项用于控制是否对次字幕应用样式覆盖
默认启用（模仿原版 mpv 渲染行为）。禁用时将和主字幕使用相同的字幕渲染逻辑

 以下为禁用--secondary-sub-override 后的次字幕渲染效果
![image](https://user-images.githubusercontent.com/61936050/196129957-f6361c8f-e6df-4987-aa90-361b1aba41cf.png)

相比使用 vf 字幕滤镜实现的次字幕的优势在于可以正确处理次字幕 PlayRes 参数和视频分辨率不匹配时的字幕渲染（次字幕显示过小）。不受视频解码方式限制
尚未实现：单独调整次字幕的 pos，这对于变相实现双语字幕比较有用。但我研究了下 mpv 的 sb_ass 发现需要重构相关代码，搞不定（

windows 上的测试版编译：[mpv-win](https://github.com/dyphire/mpv-winbuild/actions/runs/3262828005)

**Reply by dyphire** [reply]
发现vlc 4.0的双字幕功能实现的比较完善（指类似potplayer的双字幕功能

两者都能单独指定次字幕的pos，这点要强于mpv

**Reply by diana7127** [reply]
请问在mpv.net中可以禁用--secondary-sub-override达到这种显示效果吗？具体需要怎么操作呢？谢谢

**Reply by hooke007** [reply]
不关前端什么事，你要做的是换它的内核 https://github.com/dyphire/mpv-winbuild/releases

**Reply by debugzxcv** [reply]
`Error parsing option secondary-sub-override (option not found)`
https://github.com/dyphire/mpv-winbuild/releases/tag/mpv_own-2023-11-24

**Reply by dyphire** [reply]
由于 mpv 新的 pr https://github.com/mpv-player/mpv/pull/13073 实现了次字幕 pos 的控制，在它合并后就可以考虑将此更改提交 pr 给上游了。在实现次字幕 pos 的控制之前增加的次字幕样式覆盖的选项其实用场景只有弹幕，局限性太大不适合提交，现在时机成熟了。

我在考虑要不要一并实现次字幕字体及大小的单独选项并提交，其他字体选项跟随主字幕更适合

**Reply by dyphire** [reply]
经过上游的一些讨论，基于 https://github.com/mpv-player/mpv/pull/13091 再次重构了次字幕样式覆盖的实现：https://github.com/dyphire/mpv/commit/041b7a9caa152066be3a2e3d86dd1594341469da 。依然是等待上游 PR 合并后提交

**Reply by verygoodlee** [reply]
@dyphire 次字幕是不是没有输出到黑边的选项，主字幕可以通过`--sub-ass-force-margins=yes`输出到黑边，
16:9的屏幕播放2:1的视频，次字幕加载弹幕，全屏时上下黑边跑不了弹幕比较浪费屏幕空间，
我目前的解决方案是通过ffmpeg的pad滤镜填充黑边 `--vf-append=pad=aspect=16/9:x=-1:y=-1`，但是感觉这个方案太重了，而且限制了不能用纯硬解`--hwdec=auto`

**Reply by dyphire** [reply]
次字幕支持输出到黑边，统一由 sub-ass-force-margins 选项控制。弹幕之所以无法输出黑边，是因为弹幕的 ass 每条内容都有 pos 值指定了显示位置，输出到黑边只支持控制无 pos 的字幕内容。在正常 ass 字幕中一般无 pos 的都是对话或歌词内容，含有 pos 都属于特效字幕。这是 ass 标准的限制，无解

**Reply by hooke007** [reply]
如果觉得vf慢的话，你可以考虑向上游提video-pad的请求。video-crop也是不久前通过请求实现的
也可能需要额外强调一下你的用例，<s>我看之前好像video-crop之后字幕仍是按crop之前的位置渲染的</s>https://github.com/mpv-player/mpv/issues/13423

**Comment by YSYHZ** [comment]
我仔细看了上面的问答，可以将视频内置的字幕作为次字幕输出，但是脚本中没有提供对应快捷键来加载视频内置的字幕只能加载外挂字幕，不知道能否添加这个功能，感谢

**Reply by hooke007** [reply]
<s>麻烦</s>懒不是很想做。建议用替代方案（修改版mpv）或者直接敲命令

**Reply by YSYHZ** [reply]
好吧，看来只能自己敲命令了

**Comment by hooke007** [comment]
此帖终结。https://github.com/mpv-player/mpv/pull/13121 已合并
相关脚本中的hack方法也将移除

原生的实现方法 https://github.com/hooke007/MPV_lazy/wiki/0_FAQ#%E9%AB%98%E7%BA%A7%E6%AC%A1%E5%AD%97%E5%B9%95

REF https://mpv.io/manual/master/#options-secondary-sub-ass-override
